lib/common/udp_tunnel.c

```diff
@@ -0,0 +1,208 @@
+
+    h2o_buffer_dispose(&tunnel->inbuf);
+    h2o_socket_close(tunnel->sock);
+    free(tunnel);
+}
+
+static void tunnel_socket_on_read(h2o_socket_t *sock, const char *err)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = sock->data;
+
+    if (err != NULL) {
+        tunnel->super.on_read(&tunnel->super, err, NULL, 0);
+        return;
+    }
+
+    uint8_t buf[16384];
+    struct msghdr mess;
+    struct sockaddr sa;
+    struct iovec iov;
+    memset(&mess, 0, sizeof(mess));
+    mess.msg_name = &sa;
+    mess.msg_namelen = sizeof(sa);
+    iov.iov_base = buf;
+    iov.iov_len = sizeof(buf);
+    mess.msg_iov = &iov;
+    mess.msg_iovlen = 1;
+    ssize_t rret;
+    while ((rret = recvmsg(h2o_socket_get_fd(tunnel->sock), &mess, 0)) == -1 && errno == EINTR)
+        ;
+
+    h2o_socket_read_stop(tunnel->sock);
+
+    if (tunnel->super.on_udp_read) {
+        h2o_iovec_t hiov = h2o_iovec_init(iov.iov_base, iov.iov_len);
+        tunnel->super.on_udp_read(&tunnel->super, NULL, &hiov, 1);
+    } else {
+        struct {
+            uint8_t buf[8];
+            uint8_t *end;
+        } varint;
+        h2o_iovec_t vec[3];
+        const uint8_t zero = 0;
+
+        varint.end = ptls_encode_quicint(varint.buf, iov.iov_len);
+        vec[0].base = (char *)&zero;
+        vec[0].len = sizeof(zero);
+        vec[1].base = (char *)varint.buf;
+        vec[1].len = varint.end - varint.buf;
+        vec[2] = h2o_iovec_init(iov.iov_base, iov.iov_len);
+        tunnel->super.on_read(&tunnel->super, NULL, vec, sizeof(vec) / sizeof(vec[0]));
+    }
+
+    return;
+}
+
+h2o_iovec_t get_next_chunk(const uint8_t *bytes, size_t len, size_t *to_consume, int *skip)
+{
+    const uint8_t *start = bytes;
+    const uint8_t *end = bytes + len;
+    uint64_t chunk_type, chunk_length;
+
+    chunk_type = ptls_decode_quicint(&bytes, end);
+    if (chunk_type == UINT64_MAX)
+        return h2o_iovec_init(NULL, 0);
+    chunk_length = ptls_decode_quicint(&bytes, end);
+    if (chunk_length == UINT64_MAX)
+        return h2o_iovec_init(NULL, 0);
+
+    /* chunk is incomplete */
+    if (end - bytes < chunk_length)
+        return h2o_iovec_init(NULL, 0);
+
+    /*
+     * https://tools.ietf.org/html/draft-ietf-masque-connect-udp-03#section-6
+     * CONNECT-UDP Stream Chunks can be used to convey UDP payloads, by
+     * using a CONNECT-UDP Stream Chunk Type of UDP_PACKET (value 0x00).
+     */
+    *skip = chunk_type != 0;
+    *to_consume = (bytes + chunk_length) - start;
+
+    return h2o_iovec_init(bytes, chunk_length);
+}
+
+static void tunnel_on_writev(h2o_tunnel_t *_tunnel, h2o_iovec_t *iov, size_t iovlen)
+{
+    struct st_h2o_udp_tunnel_t *tunnel = (void *)_tunnel;
+    int fd;
+    fd = h2o_socket_get_fd(tunnel->sock);
+
+    struct msghdr mess;
+    struct iovec vec[iovlen];
+    for (size_t i = 0; i < iovlen; i++)
+        vec[i] = (struct iovec){
+            .iov_base = iov[i].base,
+            .iov_len = iov[i].len,
+        };
```

Comment:
I'm not sure if this is correct. IIUC, each `iov` is a UDP datagram. Assuming that is the case, I think `sendmsg` should be called for each `iov` element?

PS. And assuming that is the case, I think we should refrain from calling the arguments "iov."

Reply:
Addressed in 40e3399 (#2568).